### ISSUE TITLE
Run test suites in parallel

### ISSUE THREAD

**Issue by amolenaar** [unclebob/fitnesse#806]
How about running all tests in a suite in parallel?

**Comment by six42** [comment]
Would be a nice feature to speedup testing.
Should be optional as not all tests are build to run in parallel.
Also allow to run all suites in parallel, while the test pages are executed sequentially.

I started some work to run all testsystems in parallel. Still in an early stage. Contributions would be welcome.

**Comment by amolenaar** [comment]
I'm a little concerned on how to deal with SuiteSetUp/TearDown, as well as how it will show in the browser.

**Comment by six42** [comment]
SuiteSetUp and TD must run multiple times. Once in each parallel stream.
Display, collection and sorting of results are good questions but would expect this can be managed.
How do you want to decide how many parallel test streams are started and how are the pages/suites distributed to the different streams?
Do you plan to run in  multiple processes or threads?

**Comment by jplambert** [comment]
It can be done from outside FitNesse, by running in parallel partial suites, instead of launching once the whole suite. We could call that distributed test running. However today this is a bit clumsy as you need to split the suites by hand and maintain them thereafter, and consolidating the results into a single report isn't easy and the result isn't as nice. Furthermore, balancing the time taken by each suite to optimize the total time taken is problematic as it will be done entirely by hand.

Why am I binging this up? Because, as mentioned by @six42 , tests may not be built to run in parallel. But you can still run them in the distributed fashion I mentioned by launching them on several SUT (be it several actual computers, several virtual machines, several docker containers, or several Slim servers on the same machine). And if you have an easy way to do distributed running, then you'll have your parallel running for any tests ;-)

What could we do to improve the situation? Other testing tools like Robot Framework or Gatling allow for instance to split the report generation in two parts: (1) generate the data in XML or whatever form and (2) actually generate a report from the data. The testing tool can then be configured to do only (1) instead of (1) + (2), and then another utility tool can be used to merge several sets of data before performing (2). Thus you have a way to have a single and nice report even though you have distributed your tests.

The other big problem, then, is how to split a single test suite in several, smaller test suites. What we want ideally is for this to be done automatically, because maintaining manually such test suites will be a pain: whenever you'll add a test, you'll forget once in a while to add it to one of the distributed test suites. Or the lazy developers will add all new tests to the same suite instead of thinking it through, make it run longer and longer, and canceling the purpose of distributed running. Giving an optimal answer to this problem is not easy:
- Splitting arbitrarily the test suites may result in discrepancies in the execution time of each suites. Still maybe some dumb split (try to put the same number of test pages in each suite) might work ok enough in practice for most people. Also, try to keep together the sub-test suites that share a common suite setup, if possible.
- Or FitNesse would need to analyze the test execution time history to make a more educated guess on how to split the test suite and getting the same execution time for all distributed suites. Not sure how that would fit in FitNesse design.

In any case, that would need some way to request such a split from FitNesse. I think that something like "virtual tags" would fill the bill: just like you could split your test suite by using tags, but these tags would be dynamic and generated by FitNesse. Basically these virtual tags would convey the information "split the suite in M parts, and give me the N-th part". Maybe we could use some special symbol in these virtual tags to make sure they cannot overlap with user-defined tags?

What's your take on this? All I can say is: we would use it for sure, since we currently experience the issue of long running suites + tests that cannot be run in parallel... So far we haven't gone through the distributed path but I believe we will for sure someday. I just don't want to have to deal with the split by hand, I know it won't be maintained properly... And the reports won't be as nice. :-(

**Comment by jplambert** [comment]
Also, my understanding is that we can already have FitNesse provide only the test result data by using another responder. That's what is done by Jenkins, allowing to get FitNesse reports directly inside Jenkins.

What is missing, as far as I know, is a way to get back a full and nice report in FitNesse-style, from the data. You'll get that only when you run the tests as-is from FitNesse.

**Comment by amolenaar** [comment]
@jplambert The idea on split report generation should not have to be that hard, since each test generates it's own report file. All that's left is to merge the "suite" report files into one. Obviously we need a way to know that suite and test report files actually belong together (e.g. based on a timestamp/id)

When it comes to executing tests there are a few options:

1. Launch a SUT for each execution thread. 
2. Run multiple "service" threads on the same SUT - actions are executed in parallel on one SUT.

The first options seems simpler to implement at first and can work for both Fit and Slim suites (the Slim fixtures used for testing FitNesse maintain some global state, so multi threaded execution on the SUT would be a real challenge :) ). The first option will definitely eat up more sockets than the second.

Then there is the task of distributing the tests over the work queues. We have to take into consideration that not all tests can be executed in parallel. Tagging suites and tests seems like a good starting point to me. But even then, there are quite a few flavours to choose from:

1. parallelise  each sub-suite.
2. parallelise on tags
3. parallelise tests under a tagged page
4. build work queues based on page tags, grouping all pages with the same set of tags in one queue.

On a technical level, it should be possible to create a custom TestRunner (we now only have one: `MultipleTestsRunner`).

**Comment by fhoeben** [comment]
To what extend can running in parallel be accomplished using the jUnit runner, maven and surefire? According to https://maven.apache.org/surefire/maven-surefire-plugin/examples/fork-options-and-parallel-execution.html it should be possible to have it run tests in parallel if the runner is a child of ParentRunner (which FitNesseRunner is). 
It is an open question: I really have not looked at it yet.

**Comment by amolenaar** [comment]
I did some tests with executing FitNesse's own test set in parallel. That seems to be a challenge so far, mainly because a lot of tests start opening sockets. I want to resolve this first since it will provide some ground to test with parallel execution on the FitNesse codebase itself.

**Comment by pascaldufour** [comment]
We are starting running test in parallel. Any experience updates?. We are starting with different tags and run each tag on a separate VM. A difficult issue is how to combine the test results from 1 run which in reality are 2 fitnesse runs.  We keep searching for a solution.

**Comment by fhoeben** [comment]
I'm currently also running multiple suites in parallel, each in their own JVM. We run using separate (parallel) jobs in our build server, each having their own result and a single dashboard showing the combined status. Not ideal, but quite workable.

The main drawback, for me, is the 'static' configuration of the parallel jobs. Ideally one would be able to indicate the number of parallel executors and the tests would be distributed so no executors are ideal as long as there are 'to be executed' jobs.

**Comment by YogiKhan** [comment]
Did someone tried the solution provided by:
http://boontool.com/2011/02/18/fitnesse-trouble/

We are also in-need to run tests in parallel. Our tests are independent to each other. Any solution is welcome.

**Comment by JeremyCarlsten** [comment]
Has anyone come up with a decent solution for creating a summary report? 

I have been able to get individual tests (all browser based e2e type tests) to run in parallel at a suite level using the JunitHelper and junit's ability to run in parallel. However, the JunitHelper creates it's own JavaFormatter that seems to keep track of the test results (correct me if I'm wrong!) . I have been unsuccessful in getting access to the full result set for building the full suite summary. 

Ex. 

FrontPage.Integration  
     - SomeSuite
     - SomeOtherSuite

I can get both of the sub-suites to run in parallel; however, I don't receive the report for FrontPage.Integration as I have to run each suite independently. Perhaps something built into the framework would work better. 

Thoughts?

**Comment by YogiKhan** [comment]
I am able to run the fitnesse tests in parallel and merge the tests report and create a single report as well.

I am using jsoup to get the tests links and putting them in queue and then using threadexecutor i am able to run the tests in parallel.

For report generation, i have written a utility to merge the xml and then written xsl to get the report in html and email format.

**Comment by JeremyCarlsten** [comment]
Yeah I had considered writing a utility for it. But we've grown used to the html output of the junitHelper. Any chance yours is committed publicly?

**Comment by YogiKhan** [comment]
Its not public now but i will make it public and will share the github link.

**Comment by fhoeben** [comment]
The best reporting I've been able to do so far (when running multiple suites, each in their own JVM) has been:
* generate a surefire XML report file for each test page
* let the CI server generate a combined overview of all these test (out-of-the-box functionality of, for instance, Jenkins)
* publish the HTML report generated by the FitNesseRunner for each suite (separately, not in a single report)

This works but a single HTML overview would be better. 

I did get a request to support the [Allure](http://allure.qatools.ru/) framework. I have not really used it on large, or parallel, test sets myself but I believe it should be usable to combine the results of multiple test runs (that were executed in parallel). I did not really care for the amount of clicking I had to do to get to actual test failures. BUT its dashboard looks quite cool. 
I made a separate project to contain the [jUnit listener that exports FitNesse results to Allure](https://github.com/fhoeben/allure-fitnesse-listener).

**Comment by YogiKhan** [comment]
@jplambert here you can find the github: https://github.com/YogiKhan/fitnesse-parallel

**Comment by JeremyCarlsten** [comment]
@YogiKhan  thanks!

**Comment by fhoeben** [comment]
@YogiKhan can you me add a little explanation how to use it to the README of the repo? Thanks.

**Comment by roydekleijn** [comment]
@YogiKhan  yep, curious to see the details.

**Comment by maikelgithub** [comment]
@YogiKhan could you please add an explenation on how to use your fitnesse-parallel project?

**Comment by fhoeben** [comment]
My project now also has a [small tool to combine the HTML reports of multiple runs into a single one](https://github.com/fhoeben/hsac-fitnesse-fixtures/blob/master/src/main/java/nl/hsac/fitnesse/junit/reportmerge/HtmlReportIndexGenerator.java):
 * run multiple suites generating standard html report
 * create a single top-level root directory (e.g. target/combined-results)
 * copy the html results of each run to a child (e.g. target/combined-results/run1, target/combined-results/run2)
 * run the tool passing the top level directory (e.g. `java -cp ${FITNESSE_CP} nl.hsac.fitnesse.junit.reportmerge.HtmlReportIndexGenerator target/combined-results`)

It will create an index.html in the top level directory (e.g. target/combined-results/index.html)  providing an overview (and highlighting any failures/errors), and a CSV and JSON summary for automated processing

![screenshot 2019-03-07 at 17 04 04](https://user-images.githubusercontent.com/5467859/53970483-1d0ef800-40fb-11e9-8cc5-8fe51a5905dc.png)

**Comment by krishabh85** [comment]
@YogiKhan  It is a bit difficult to understand your work but it seems that it will work absolutely good. Can you add more explanation and examples to make us more clear. It will help me in solving one major task for my company. Kindly, help.
Thanks
Rishabh Khandelwal

**Comment by dinesh-01** [comment]
Currently i am achieving this by running fitnesse in docker  based on tag and jenkinsFile
example :- mvn -Pfitnesse test -Dfitnesse.command='-c FrontPage.ExamplePage?suite&suiteFilter=$tag_name&format=xml&includehtml' -Dfitnesse.result='fitnesse-data-result.xml'

My POM file :-  <echo taskname="fitnesse" message="Starting FitNesse at http://localhost:${fitnesse.port}"/>
                  <java classname="fitnesseMain.FitNesseMain" classpathref="maven.runtime.classpath" fork="true">
                    <arg line="-p ${fitnesse.port}" />
                    <arg line="-d ." />
                    <arg line="-l ." />
                    <arg line="-b ${fitnesse.result}" />
                    <arg line="${fitnesse.command}" />

**Comment by violinner** [comment]
Is there an existing tool that will combine several XML FitNesse reports into one?
We are running parallel tests in a Jenkins pipeline, and the Jenkins FitNesse history report plugin is not happy with multiple fitnesse-results.xml files.

**Comment by fhoeben** [comment]
@violinner I don't believe there is one yet...

**Comment by YogiKhan** [comment]
> Is there an existing tool that will combine several XML FitNesse reports into one?
> We are running parallel tests in a Jenkins pipeline, and the Jenkins FitNesse history report plugin is not happy with multiple fitnesse-results.xml files.

That’s exactly the problem I faced when we ran tests in parallel. So I had wrote my own utility to combine the results and publish as HTML to Jenkins and email as xml

**Comment by violinner** [comment]
Can you share that code?  I cannot use the HTML version, but if the files are merged into one with an accurate grand total, that would fix my current pipeline upgrade dilemma.

**Comment by fhoeben** [comment]
@jplambert have you seen [my PR to spit the tests](https://github.com/unclebob/fitnesse/pull/1268) as you described in your initial comment 

> In any case, that would need some way to request such a split from FitNesse. I think that something like "virtual tags" would fill the bill: just like you could split your test suite by using tags, but these tags would be dynamic and generated by FitNesse. Basically these virtual tags would convey the information "split the suite in M parts, and give me the N-th part". Maybe we could use some special symbol in these virtual tags to make sure they cannot overlap with user-defined tags?

I also created my own tool to combine the html reports (the surefire xml reports can just be dumped in one directory and will all be picked up). 

HtmlReportIndexGenerator is also available as standalone docker image https://hub.docker.com/r/hsac/fitnesse-fixtures-combine

**Comment by tcnh** [comment]
I just use the allure listener and aggregate the allure results in a step after parallel execution. Then run the allure report geneartion afterwards using the combined allure-results. Gives me a nice complete dashboard of the combined parallel runs as if it were only one test run

**Comment by violinner** [comment]
I'll check it out, thanks.

**Comment by YogiKhan** [comment]
> I'll check it out, thanks.

Here is the link https://github.com/YogiKhan/fitnesse-parallel

**Comment by violinner** [comment]
Combining fitnesse-report.xml files:
https://github.com/YogiKhan/fitnesse-parallel/blob/master/fitnesse-parallel/src/main/java/com.fitnesse.parallel/XMLConcat.java
Bingo.

**Comment by YogiKhan** [comment]
> Combining fitnesse-report.xml files:
> https://github.com/YogiKhan/fitnesse-parallel/blob/master/fitnesse-parallel/src/main/java/com.fitnesse.parallel/XMLConcat.java
> Bingo.

Let me know if you have any questions in running that.

**Comment by YogiKhan** [comment]
> @YogiKhan It is a bit difficult to understand your work but it seems that it will work absolutely good. Can you add more explanation and examples to make us more clear. It will help me in solving one major task for my company. Kindly, help.
> Thanks
> Rishabh Khandelwal

I did added some more explanation.Let me know if you are still looking for help.

**Comment by fhoeben** [comment]
@YogiKhan maybe some additional text in the README describing what the entire project does (in parallel) and also specifically what the two parameters do (now it just says one can enter high values, but what do they indicate/control?)

**Comment by violinner** [comment]
@YogiKhan 
We are running FitNesse through a Jenkins pipeline, and summarizing the results with the Jenkins FitNesse plug-in.  The combined right/wrong/skipped/exception figures, formerly in a testResults/finalCounts element, are now in each sub-suite result that was concatenated.

The consequence for this is that while the FitNesse History in Jenkins is now consistent with non-parallel execution of the same suites, the "FitNesse Results Trend" graph is now showing each Suite run in parallel as one test, clearly incorrect.

I will try an experiment, and add a top-level element, "finalCounts," with the total of all the ones in all the parallel results, to see if the Jenkins FitNesse plug-in will behave again.

**Comment by violinner** [comment]
Update: The Jenkins FitNesse plug-in reads the test results of nested fitnesse XML-formatted output, but the grand totals are not available for the summary trend graph.  Adding a free-standing "finalCounts" element does not change the results.
